src/gridstack.ts

```diff
@@ -2146,11 +2146,7 @@ export class GridStack {
         this.engine.addedNodes.push(node);// @ts-ignore
         this._triggerAddEvent();// @ts-ignore
         this._triggerChangeEvent();
-
         this.engine.endUpdate();
-        if (this._gsEventHandler['dropped']) {

```

Comment:
callback here should work as line 2157 has comment about checking that we are here (and not deleted by user) so I need to better understand what crash you were seeing - calling after that statement doesn't make sense of that statement so something else needs to change as 2158 will always be true.

conceptually I don't see why CB can't be called after like you have, other than maybe doing all that work preparing a node that won't exists in the scene soon after...

Reply:
Hi, thnx for the fast reply,

The situation can be described as the next in pseudo code:

```
protected items: [item1, item2, item3]

handleDropped() {
	this.items = [];
}

<gridstack id=“grid1”>
	<gridstack-item *ngFor="let item of items;
		<myItem item=“item” />
	</gridstack-item>
<gridstack />

<gridstack id=“grid2” (droppedCB)="handleDropped()"><gridstack />
```

Drag an item from grid 1 to grid 2:

```
Runtime error:
core.mjs:10183 ERROR TypeError: Cannot read properties of undefined (reading 'grid')
    at dd-gridstack.js:53:51
    at Array.forEach (<anonymous>)
    at DDGridStack.draggable (dd-gridstack.js:42:33)
    at GridStack._prepareDragDropByNode (gridstack.js:1982:16)
    at gridstack.js:1868:26
    at timer (zone.js:2367:41)
    at _ZoneDelegate.invokeTask (zone.js:402:31)
    at core.mjs:25945:55
    at AsyncStackTaggingZoneSpec.onInvokeTask (core.mjs:25945:36)
    at _ZoneDelegate.invokeTask (zone.js:401:60)
```

`_prepareDragDropByNode` is called after the `(droppedCB)` already fired the event signalling that it was complete. We cleared grid 1 immediately on the event. This interferes with the code in the `setTimeout`

If we emit the ‘dropped’ event in the `setTimeout` like in my PR we don’t get the runtime error and all works fine. Would save us some unneeded setTimeouts in our code and would make the lib more robust. Logic wise it makes (to me) more sense to emit a done event like ‘dropped’ when something is really done and not before the process of cleaning up.

I don't have enough knowledge on  the lib to see why the setTimeout is needed in the first place.

thnx!

Reply:
Hi @adumesny,

I wondered if i can help out further to resolve this issue?  Do you need more information? Or maybe you can propose another way to fix this? I would like to help, but not sure what to do. We are using public api's so the crash should not happen. We could fork but that is definitely something we would like to avoid. We rather would like to help out in the official repo.

thnx!

Reply:
yes let's defintely fix it in the official repo - I was on vacation this week so catching up to a ton of work.

Reply:
| We cleared grid 1 immediately on the event
technically you clear the array right away, but ng needs next cycle to acutally update the dom ngFor loop (unless you also do that synchronously).

let me test why we have a timeout at all (comment doesn't have bug number and can't recall).

could you create a simple reproduceable case from the speudo code above so I can make sure to test that as well ?